internal/controllers/machine_controller.go

```diff
@@ -716,13 +746,25 @@ func (r *MachineReconciler) domainFor(
 
 	volumeStates, err := r.attachDetachVolumes(ctx, log, machine, attacher)
 	if err != nil {
+		if eventErr := r.eventStore.AddEvent(machine.Metadata, corev1.EventTypeWarning, "AttchDetachVolume", fmt.Sprintf("Volume attach/detach failed with error: %s .", err)); eventErr != nil {
+			log.Error(eventErr, "failed to add machine event")
+		}
 		return nil, nil, nil, err
 	}
+	if eventErr := r.eventStore.AddEvent(machine.Metadata, corev1.EventTypeNormal, "AttchedVolume", "Successfully attached volumes."); eventErr != nil {
+		log.Error(eventErr, "failed to add machine event")
+	}
 
 	nicStates, err := r.setDomainNetworkInterfaces(ctx, machine, domainDesc)
 	if err != nil {
+		if eventErr := r.eventStore.AddEvent(machine.Metadata, corev1.EventTypeWarning, "AttchDetachNIC", fmt.Sprintf("Setting domain network interface failed with error: %s .", err)); eventErr != nil {
+			log.Error(eventErr, "failed to add machine event")
+		}
 		return nil, nil, nil, err
 	}
+	if eventErr := r.eventStore.AddEvent(machine.Metadata, corev1.EventTypeNormal, "AttchedNIC", "Successfully attached network interfaces."); eventErr != nil {
```

Comment:
if machine doesn't have network interface, will this event generate too?

Reply:
Changed it so that it will only be printed when networkInterface is not nil. The same applies to volumes.